collector/pipelinerun.go

```diff
@@ -2,27 +2,57 @@ package collector
 
 import (
 	"context"
-	"github.com/go-kit/log/level"
+	"fmt"
+
 	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
-	"os"
 
-	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
-	"sigs.k8s.io/controller-runtime/pkg/metrics"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 )
 
-func SetupPipelineRunCachingClient(mgr ctrl.Manager) {
-	// we do not set up a reconciler to get a caching client, but we make one list
-	// call up top to get the ball rolling; confirmed via break points in a debugger that
-	// k8s methods for creating a shared informer and watch are called
-	prList := v1beta1.PipelineRunList{}
-	opts := &client.ListOptions{Namespace: metav1.NamespaceAll}
-	mgr.GetClient().List(context.Background(), &prList, opts)
-	psCollector, err := NewCollector(logger, mgr.GetClient())
+type ReconcilePipelineRun struct {
+	client        client.Client
+	scheme        *runtime.Scheme
+	eventRecorder record.EventRecorder
+	psCollector   *PipelineServiceCollector
+}
+
+func SetupPipelineRunCachingClient(mgr ctrl.Manager) error {
+	reconciler := &ReconcilePipelineRun{
+		client:        mgr.GetClient(),
+		scheme:        mgr.GetScheme(),
+		eventRecorder: mgr.GetEventRecorderFor("MetricExporterPipelineRuns"),
+		psCollector:   NewCollector(),
+	}
+	return ctrl.NewControllerManagedBy(mgr).For(&v1beta1.PipelineRun{}).Complete(reconciler)
+}
+
+func (r *ReconcilePipelineRun) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	var cancel context.CancelFunc
+	ctx, cancel = context.WithCancel(ctx)
+	defer cancel()
+	log := ctrl.Log.WithName("metricsexporter").WithValues("namespace", request.Namespace, "resource", request.Name)
+
+	pr := &v1beta1.PipelineRun{}
```

Comment:
if you mean put a TODO comment, then sure I can do that

or I can do fetches with both types, and see which comes back without an error ... I can try that, see if it works

but I can't just start using just v1 until RHTAP switches

Reply:
can't reconcile on both v1 and v1beta1 with one reconciler:

```
2023-06-22T21:03:31Z	ERROR	main	unable to start controller-runtime manager	{"error": "For(...) should only be called once, could not assign multiple objects for reconciliation"}
main.main
	/workspace/main.go:85
runtime.main
	/usr/local/go/src/runtime/proc.go:250
```

and since the dynamic client will use the version field IIRC, I just put a TODO comment that we need to change at some point to v1